pkg/server/apiextensions.go

```diff
@@ -187,106 +282,118 @@ func (c *inheritanceCRDLister) GetWithContext(ctx context.Context, name string)
 	}
 
-	// If we found the CRD in ctx's logical cluster, that takes priority.
-	if crd != nil {
-		return crd, nil
-	}
+	systemCRDKeys := c.systemCRDProvider.Keys(ctx, org, workspace)
 
-	// ClusterWorkspace CRD is apparently not installed
-	if c.workspaceLister == nil {
-		return nil, apierrors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
+	systemCRDKeyName := clusters.ToClusterAwareKey(SystemCRDLogicalCluster, name)
+	if systemCRDKeys.Has(systemCRDKeyName) {
+		crd, err = c.crdLister.Get(systemCRDKeyName)
+		if err != nil && !apierrors.IsNotFound(err) {
+			// something went wrong w/the lister - could only happen if meta.Accessor() fails on an item in the store.
+			return nil, err
+		}
+
+		if crd != nil {
+			return crd, nil
+		}
+
+		return nil, apierrors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, name)
 	}
 
-	if !helper.HasClusterWorkspace(org, ws) {
-		// Without a cluster workspace no inheritance is possible
-		return nil, apierrors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
+	// Priority 1: see if it comes from any APIBindings
+	parts := strings.SplitN(name, ".", 2)
+	resource, group := parts[0], parts[1]
+
+	// TODO(ncdc): APIResourceSchema requires that its group be "" for the core group,
+	// See if we can unify things.
+	if group == "core" {
+		group = ""
 	}
 
-	// Check for API inheritance
-	targetWorkspaceKey := helper.WorkspaceKey(org, ws)
-	workspace, err := c.workspaceLister.Get(targetWorkspaceKey)
+	apiBindings, err := c.apiBindingLister.ListWithContext(ctx, labels.Everything())
 	if err != nil {
-		// If we're here it means ctx's logical cluster doesn't have the CRD and there isn't a
-		// ClusterWorkspace for the logical cluster. Just return not-found.
-		if apierrors.IsNotFound(err) {
-			return nil, apierrors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
-		}
-
 		return nil, err
 	}
+	for _, apiBinding := range apiBindings {
+		if apiBinding.ClusterName != cluster.Name {
+			continue
+		}
+		if apiBinding.Status.Phase != apisv1alpha1.APIBindingPhaseBound {
+			// TODO(ncdc): what if it's in the middle of rebinding and it takes some time - we won't include any
+			// CRDs that were previously bound...
+			continue
+		}
 
-	if workspace.Spec.InheritFrom == "" {
-		// If we're here it means ctx's logical cluster doesn't have the CRD, the ClusterWorkspace exists,
-		// but it's not inheriting. Just return not-found.
-		return nil, apierrors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
-	}
+		for _, boundResource := range apiBinding.Status.BoundResources {
+			if boundResource.Group == group && boundResource.Resource == resource {
+				crdKey := clusters.ToClusterAwareKey("system:bound-crds", boundResource.Schema.UID)
+				crd, err = c.crdLister.Get(crdKey)
+				if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
(No action required) This pattern of error handling seems strange to me. If the only goal is returning a different error when NotFound, maybe make that explicit?

```golang
err := get(...)
if apierrors.IsNotFound(err) {
  return nil, <unavailable error>
} 
// Either crd or err will be non-nil.  No further logic required to get desired return value.
return crd, err
```

Reply:
fixed